platform/qemu/virt/config.mk

```diff
@@ -9,8 +9,8 @@
 
 # Compiler flags
 platform-cppflags-y =
-platform-cflags-y =-mabi=lp64 -march=rv64imafdc -mcmodel=medany
-platform-asflags-y =-mabi=lp64 -march=rv64imafdc -mcmodel=medany
+platform-cflags-y =-mcmodel=medany
+platform-asflags-y =-mcmodel=medany
```

Comment:
The OpenSBI is RISC-V specific and generic library under lib/ directory expects 'I', 'M', 'A', 'F', and 'D' extension. The OpenSBI dependency on 'C' extension is optional.

I think having explicit -mabi, -march, and -mcmodel is good because it ensures toolchain being used has required features and code does not use any other RISC-V extension.

Instead this change, let's update config.mk of all platforms and move -mabi, -march, and -mcmodel in top-level Makefile. The top-level Makefile will give two options PLATFORM_RISCV_XLEN and PLATFORM_RISCV_C_EXT.

If PLATFORM_RISCV_XLEN=32 then -mabi=lp32 -march=rv32imafd -mcmodel=medany
If PLATFORM_RISCV_XLEN=64 then -mabi=lp64 -march=rv64imafd -mcmodel=medany

If PLATFORM_HAS_C_EXT=y then "c" is appended to -march

By default, PLATFORM_RISCV_XLEN=64 and PLATFORM_RISCV_C_EXT=y (if not set).

Both PLATFORM_RISCV_XLEN and PLATFORM_RISCV_C_EXT can be set in config.mk or passed as command-line parameter to top-level makefile.

Suggestions???

Reply:
This is not dynamic so the virt machine won't build for 32-bit RISC-V without code modifications. Whatever we do needs to be dynamic and adjust based on the compiler.

Reply:
Okay, how about this.

When PLATFORM_RISCV_XLEN and PLATFORM_RISCV_C_EXT are not defined by platform config.mk, we set these variable in Makefile by looking at toolchain capabilities.

This way platforms have a way to enforce particular PLATFORM_RISCV_XLEN and PLATFORM_RISCV_C_EXT. If platforms such as QEMU virt/sifive_u don't want to enforce then top-level Makefile do best guess based on toolchain capabilities.

Reply:
That is more or less what we do now. If the platform hard codes the abi and arch then we will use that, if not it is up to the compiler. I can make it more modular but it will still be the same effect we have now.